src/ufuncs.c

```diff
@@ -1942,6 +1942,70 @@ static void polygonize_full_func(char** args, npy_intp* dimensions, npy_intp* st
 }
 static PyUFuncGenericFunction polygonize_full_funcs[1] = {&polygonize_full_func};
 
+static char nearest_points_dtypes[3] = {NPY_OBJECT, NPY_OBJECT, NPY_OBJECT};
+static void nearest_points_func(char** args, npy_intp* dimensions, npy_intp* steps,
+                                void* data) {
+  GEOSGeometry* in1 = NULL;
+  GEOSGeometry* in2 = NULL;
+  GEOSGeometry** geom_arr;
+  GEOSCoordSequence* coord_seq = NULL;
+
+  CHECK_NO_INPLACE_OUTPUT(2);
+
+  // allocate a temporary array to store output GEOSGeometry objects
+  geom_arr = malloc(sizeof(void*) * dimensions[0]);
+  CHECK_ALLOC(geom_arr);
+
+  GEOS_INIT_THREADS;
+
+  BINARY_LOOP {
+    /* get the geometries: return on error */
+    if (!get_geom(*(GeometryObject**)ip1, &in1)) {
+      errstate = PGERR_NOT_A_GEOMETRY;
+      destroy_geom_arr(ctx, geom_arr, i - 1);
+      break;
+    }
+    if (!get_geom(*(GeometryObject**)ip2, &in2)) {
+      errstate = PGERR_NOT_A_GEOMETRY;
+      destroy_geom_arr(ctx, geom_arr, i - 1);
+      break;
+    }
+
+    if ((in1 == NULL) | (in2 == NULL)) {
+      // in case of a missing value: return NULL (None)
+      geom_arr[i] = NULL;
+    } else if (GEOSisEmpty_r(ctx, in1) | GEOSisEmpty_r(ctx, in2)) {
+      // GEOSNearestPoints_r would otherwise return a NULL not distinguishable
+      // from an actual error
+      geom_arr[i] = NULL;
```

Comment:
An implementation aspect: if we decide to return a LineString (as I did here), this is actually a `YY_Y` type of function, and I could write a small wrapper around `GEOSNearestPoints_r` that returns the LineString, so we can use it with the `YY_Y_func` template. 

But, one difference with the other YY_Y functions (intersection, difference, union, etc) is that `GEOSNearestPoints_r` also returns NULL if any of the inputs are empty geometries (it checks for that explicitly in the C API wrapper: https://github.com/libgeos/geos/blob/cd0ade60ab050cf70c66f8c555701717e4486461/capi/geos_ts_c.cpp#L865-L866), while it in general uses a NULL return value to indicate an error (as the other functions used in YY_Y). 

Personally, I think it would be nice to not error for empty input, but handle it like we already handle None input. In addition, the GEOS exception is also completely unclear that it comes from empty geometries (although this is something that could be fixed in a future GEOS version)